kitsune/sumo/static/sumo/js/switching-devices-wizard-manager.js

```diff
@@ -8,9 +8,113 @@
  */
 export default class SwitchingDevicesWizardManager {
   #formWizard = null;
-  #state = {};
+  #fxaRoot = null;
+  #state = {
+    fxa: {
+      // These UTM parameters are fallback values in the event that we
+      // don't get any passed through to us from the entrypoint to this
+      // SUMO page. These are what will ultimately get passed through the
+      // FxA sign-in flow.
+      utm_source: "support.mozilla.org",
+      utm_campaign: "migration",
+      utm_medium: "mozilla-websites",
+
+      entrypoint: "fx-new-device-sync",
+      entrypoint_experiment: null,
+      entrypoint_variation: null,
+      flow_id: null,
+      flow_begin_time: null,
+      context: null,
+    },
+  };
 
   constructor(formWizard) {
     this.#formWizard = formWizard;
+
+    if (!this.#formWizard.hasAttribute("fxa-root")) {
+      throw new Error(
+        "Expected fxa-root on element passed to " +
+          "SwitchingDevicesWizardManager constructor."
+      );
+    }
+    this.#fxaRoot = this.#formWizard.getAttribute("fxa-root");
+
+    // If the page was loaded with any UTM or entrypoint parameters, we'll want
+    // to pass those along, so now we check for those and add them to our state
+    // if need be. Since this occurs during construction, we don't need to call
+    // #updateState().
+    let searchParams = new URLSearchParams(window.location.search);
+
+    for (let paramName of [
+      "utm_source",
+      "utm_campaign",
+      "utm_medium",
+      "entrypoint",
+      "entrypoint_experiment",
+      "entrypoint_variation",
+    ]) {
+      if (searchParams.has(paramName)) {
+        this.#state.fxa[paramName] = searchParams.get(paramName);
+      }
+    }
+
+    // We need to get some query parameters from the FxA server before we
+    // show the user any kind of form to create or sign-in to an account.
+    // See https://mozilla.github.io/ecosystem-platform/relying-parties/reference/metrics-for-relying-parties#relying-party-hosted-email-form.
+    this.#requestMetricsParams();
+  }
+
+  /**
+   * Returns a fresh copy of the internal state object. This is mainly
+   * for testing.
+   *
+   * @type {object}
+   */
+  get state() {
+    return Object.assign({}, this.#state);
+  }
+
+  /**
+   * Updates the internal state object by replacing it with a copy of newState.
+   *
+   * @param {object}
+   *   The new state object.
+   */
+  #updateState(newState) {
+    this.#state = Object.assign({}, newState);
+  }
```

Comment:
```suggestion
  #updateState(newState) {
    this.#state = Object.assign(#this.state, newState);
  }
```

Might be nice to be able to pass in an objecting representing the diff, rather than having to always clone `this.#state` before modifying it